### PR TITLE
Auth Method Insight

### PR DESCRIPTION
**Pull request by lordkyzr**
This PR is missing integration tests still but I am looking for feedback on the code itself to make sure everything is copasetic.

### CODE REVIEW

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs [hunk 1]
```diff
@@ -38,6 +40,8 @@ public void Constructor_ShouldSetProperties()
             Assert.AreEqual(true, o.Response);
             Assert.AreEqual("deviceId", o.DeviceId);
             Assert.AreSame(devicePins, o.DevicePins);
+            Assert.AreEqual(null, o.AuthPolicy);
+            Assert.AreEqual(null, o.AuthMethods);
         }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs]: Where are the tests for when the values are not null?

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs [hunk 2]
```diff
@@ -41,6 +43,8 @@ public void Constructor_ShouldSetProperties()
             Assert.AreEqual("Type", o.Type);
             Assert.AreEqual("Reason", o.Reason);
             Assert.AreEqual("Denial Reason", o.DenialReason);
+            Assert.AreEqual(null, o.AuthPolicy);
+            Assert.AreEqual(null, o.AuthMethods);
         }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs]: Where are the tests for when the values are not null?

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    /// Represents an authentication method in an authentication response
+    /// </summary>
+    public class AuthMethod
+    {
+        /// <summary>
+        /// Represents the type of method
+        /// </summary>
+
+        //MAKE THIS AN ENUM
+        public AuthMethodType Method { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+
+        /// <summary>
+        /// Whether the subscriber has set up this auth method
+        /// </summary>
+        public bool? Set { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether this auth method can be used during a request
+        /// </summary>
+        public bool Active { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether the subscriber has explicitly disallowed this auth method
+        /// </summary>
+        public bool Allowed { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether the auth method is supported on the subscriber's device
+        /// </summary>
+        public bool Supported { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether the subscriber requires this auth method with every request
+        /// </summary>
+        public bool? UserRequired { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether the method was successfully verified or not
+        /// </summary>
+        public bool? Passed { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+        /// <summary>
+        /// Whether there was an error preventing verification of this auth method
+        /// </summary>
+        public bool? Error { get; set; }
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Should only be get if you are setting it in the constructor

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs (hunk 2 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs]: I have added another test to test when the values are not null

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs (hunk 1 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs]: I have added another test to test when the values are not null

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs (hunk 3 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: This has been updated.

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    /// <summary>
+    /// Represents an authentication method in an authentication response
+    /// </summary>
+    public class AuthMethod
+    {
+        /// <summary>
+        /// Represents the type of method
+        /// </summary>
+
+        //MAKE THIS AN ENUM
```
[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk/Domain/Service/AuthMethod.cs]: Removed

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs [hunk 12]
```diff
@@ -333,15 +345,48 @@ public void HandleWebhook_AuthPackage_ShouldReturnProperObject()
             Assert.AreEqual(Sdk.Domain.Service.AuthorizationResponseType.AUTHORIZED, authResponse.Type);
             Assert.AreEqual(Sdk.Domain.Service.AuthorizationResponseReason.APPROVED, authResponse.Reason);
             Assert.AreEqual("Denial Reason", authResponse.DenialReason);
-            Assert.AreEqual(false, authResponse.Fraud);
+            Assert.AreEqual(false, authResponse.Fraud);
+            Assert.AreEqual(2, authResponse.AuthPolicy.RequiredFactors);
+            Console.WriteLine(authResponse.AuthPolicy.Locations[0].Radius);
```
[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: Removed

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs [hunk 13]
```diff
@@ -17,11 +17,12 @@ public DirectorySessionSteps(CommonContext commonContext, DirectoryClientContext
             _directoryClientContext = directoryClientContext;
         }
 
-        [Given(@"I made a Device linking request")]
-        public void GivenIMadeADeviceLinkingRequest()
-        {
-            _directoryClientContext.LinkDevice(Util.UniqueUserName());
-        }
+        //[When(@"I make a Device linking request")]
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs]: Why the commented code? If it's not used, it should be deleted.

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs [hunk 14]
```diff
@@ -66,5 +67,79 @@ public void WhenIAttemptToRetrieveTheSessionListForTheUser(string p0)
...
+                var deviceStatus = device.Status;
+
+                if (device.Status.StatusCode != 1)
+                {
+                    throw new System.Exception($"All Devices should be linked. Device status was {device.Status.Text}");
+                }
+            }
+        }
+
+        //[Then(@"all of the devices should be active during timeout")]
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs]: More commented code

**Comment by lordkyzr** [comment]
This PR has been taken out of draft with the Appium integration tests being completed and feedback has been incorporated and is ready for full review.

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs (hunk 13 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs]: Removed

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs (hunk 14 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/DirectorySessionSteps.cs]: Removed

**Comment by aenglander** [comment]
Looks like you need to rebase as this cannot be merged.

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs [hunk 15]
```diff
@@ -393,5 +438,1008 @@ public void HandleWebhook_ShouldThrowOnUnexpectedResult()
...
+            mockTransport.Verify();
+
+            Assert.AreEqual(true, response.AuthPolicy.RequireInherenceFactor);
+            Assert.AreEqual(false, response.AuthPolicy.RequireKnowledgeFactor);
+            Assert.AreEqual(false, response.AuthPolicy.RequirePosessionFactor);
+            Assert.AreEqual(null, response.AuthPolicy.RequiredFactors);
+        }
+
+        [TestMethod]
+        public void AuthMethodInsight_InvalidTypeShouldNotFail()
```
[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: Yes in the Python SDK it is using warnings to inform the user. In dotnet the equivalent looks like its WarningException but it appears it does raise a full exception to the implementer unless they catch it.

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs [hunk 16]
```diff
@@ -393,5 +438,1008 @@ public void HandleWebhook_ShouldThrowOnUnexpectedResult()
...
+            authMethod.Supported = supported;
+            authMethod.UserRequired = userRequired;
+            authMethod.Passed = passed;
+            authMethod.Error = error;
+            return authMethod;
+
+        }
+
+        [TestMethod]
+        public void AuthMethodInsight_1_UnknownAuthMethodShouldNotThrowError()
```
[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: The AuthMethod objects are built when the JSON is parsed from the response. The CreateAuthTransportMethod function was just a helper function I wrote for the tests so I didn't have to copy paste the creation of the object over and over. I know all of the JSON parsing is tested elsewhere so do we just want to remove these tests?

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (bporter125) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: @aenglander We're using the warnings module like with our other warns. So correct it's not just a print.

```
warnings.warn(
                            "Invalid policy type given: %s. "
                            "It will be ignored, but this could "
                            "signify the need for an update." % type)
```

Do you have any input for Branden regarding how we could issue a warning? Raising an exception sounds more drastic than we should be.

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (bporter125) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: If you are saying we have tests that properly tests the various inputs we can receive, then I'm fine with removing them. This would include verifying each of the enums we give below can be parsed, as well as that each respective attribute can accept true, false, and null.

[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: After our discussion it appears we are covering the test cases you wanted to make sure were covered.  We are going to leave the tests in for now since they are testing the underlying object with the caveat that there is a risk to future maintenance plans in keeping them.

File: src/iovation.LaunchKey.Sdk.Tests.Integration/Features/directory_client/devices/directory-client-device-link.feature [hunk 17]
```diff
@@ -20,3 +20,13 @@ Feature: Directory Client can link Devices
     When I make a Device linking request
     And I retrieve the Devices list for the current User
     Then there should be 1 Device in the Devices list
+    And all of the devices should be inactive
+    
+  @device_testing
```
Comment (bporter125) [inline src/iovation.LaunchKey.Sdk.Tests.Integration/Features/directory_client/devices/directory-client-device-link.feature]: We should come up with a way to prevent these tests from running if Appium or Kobiton aren't configured. Similar to: https://github.com/iovation/launchkey-python/blob/3.5-DEV/features/base_environment.py#L110

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthorizationResponse.cs [hunk 18]
```diff
@@ -62,7 +62,17 @@ public class AuthorizationResponse
...
+        /// The criteria the user had to fulfill to authorize the request successfully
+        /// </summary>
+        public AuthPolicy AuthPolicy { get; }
+
+        /// <summary>
+        /// A list of all Auth Methods and the role they played in this Authorization request
+        /// </summary>
+        public IList<AuthMethod> AuthMethods { get; }
+
+        public AuthorizationResponse(string authorizationRequestId, bool authorized, string serviceUserHash, string organizationUserHash, string userPushId, string deviceId, List<string> devicePins, AuthorizationResponseType? type, AuthorizationResponseReason? reason, string denialReason, bool? fraud, AuthPolicy authPolicy, IList<AuthMethod> authMethods)
```
Comment (bporter125) [inline src/iovation.LaunchKey.Sdk/Domain/Service/AuthorizationResponse.cs]: Do we have tests to verify that authPolicy and authMethods are nullable? Maybe not in this object, but at the very least further up we need to expect the device response to not contain both of these fields. Similar to: https://github.com/iovation/launchkey-python/blob/3.5-DEV/tests/test_entities_service.py#L278 and https://github.com/iovation/launchkey-python/blob/3.5-DEV/tests/test_entities_service.py#L449.

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (aenglander) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: Let’s go with warnings in the trace subsystem. https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.tracewarning

Let’s also make sure to update the documentation to state that we use the trace subsystem for warnings.

[the remark this answers is not in the document]
Comment (aenglander) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: Let’s go with warnings in the trace subsystem. https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.tracewarning

Let’s also make sure to update the documentation to state that we use the trace subsystem for warnings.

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/OrganizationDirectorySteps.cs [hunk 19]
```diff
@@ -257,6 +257,7 @@ public void WhenIAttemptRetrieveTheDirectoryIdentifiedBy(string directoryId)
             }
         }
 
+        [Given("I have added an SDK Key to the Directory")]
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/OrganizationDirectorySteps.cs]: Missing `@`

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs [hunk 20]
```diff
@@ -7,8 +8,83 @@ namespace iovation.LaunchKey.Sdk.Tests.Transport.Domain
...
+            var authorizationRequestId = Guid.NewGuid();
+            var devicePins = new[] { "1", "2" };
+
+            var authPolicy = new AuthPolicy.JWEAuthPolicy
+            {
+                Requirement = "types",
+                Types = new List<string> { "possession" }
+            };
+
+            var authMethods = new AuthPolicy.AuthMethod[7];
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs]: Probably cleaner to initialize in-line: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/single-dimensional-arrays#array-initialization

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs [hunk 21]
```diff
@@ -7,13 +8,47 @@ namespace iovation.LaunchKey.Sdk.Tests.Transport.Domain
...
             var devicePins = new[] { "1", "2" };
+
+            var authPolicy = new AuthPolicy.JWEAuthPolicy
+            {
+                Requirement = "types",
+                Amount = 3,
+                Types = new List<string> { "inherence" }
+            };
+
+            var authMethods = new AuthPolicy.AuthMethod[7];
```
Comment (aenglander) [inline src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs]: See earlier comment about initialization

File: src/iovation.LaunchKey.Sdk.Tests.Integration/Features/directory_client/devices/directory-client-device-link.feature (hunk 17 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests.Integration/Features/directory_client/devices/directory-client-device-link.feature]: This has been completed

File: src/iovation.LaunchKey.Sdk/Domain/Service/AuthorizationResponse.cs (hunk 18 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk/Domain/Service/AuthorizationResponse.cs]: I think I found the corresponding tests and I think we have that coverage. Can you verify the tests (here)[https://github.com/iovation/launchkey-dotnet/blob/c8885909f844db7b459051265b5febbdfbae044c/src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseDeviceJWETests.cs#L283] and (here)[https://github.com/iovation/launchkey-dotnet/blob/c8885909f844db7b459051265b5febbdfbae044c/src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseDeviceJWETests.cs#L89] and see if that meets what you are looking for?

Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk/Domain/Service/AuthorizationResponse.cs]: If not I am better able to trace through the different object layers and show you what we have better than I was able to the other day.

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs (hunk 20 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs]: So you are thinking 
```
            var authMethods = new AuthPolicy.AuthMethod[7]
            {
                CreateAuthTransportMethod("wearables", false, false, true, true, null, null, null),
                CreateAuthTransportMethod("geofencing", null, true, true, true, null, null, null),
                CreateAuthTransportMethod("locations", true, true, true, true, true, false, false),
                CreateAuthTransportMethod("pin_code", true, true, true, true, true, null, null),
                CreateAuthTransportMethod("circle_code", false, false, true, true, null, null, null),
                CreateAuthTransportMethod("face", false, false, true, true, null, null, null),
                CreateAuthTransportMethod("fingerprint", false, false, true, true, null, null, null)
            };
```

File: src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/OrganizationDirectorySteps.cs (hunk 19 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests.Integration/SpecFlow/Steps/OrganizationDirectorySteps.cs]: Fixed

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs (hunk 20 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServerSentEventAuthorizationResponseTests.cs]: This is fixed

File: src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs (hunk 21 above)
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Transport/Domain/ServiceV3AuthsGetResponseTests.cs]: This is also fixed

File: src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (lordkyzr) [reply src/iovation.LaunchKey.Sdk.Tests/Client/BasicServiceClientTests.cs]: I have uploaded the changes so we are sending a TraceWarning if the policy type is invalid. This should be resolved.
